data_pipeline/gmail_data/get_gmail_data.py

```diff
@@ -46,14 +52,94 @@ def iter_link_message_thread(service: Resource, user_id: str) -> Iterable[dict]:
+    df_link = pd.DataFrame(iter_link_message_thread_ids(service, user_id))
+    df_link['user_id'] = user_id
+    df_link['imported_timestamp'] = imported_timestamp
     return df_link
 
 
+@backoff.on_exception(backoff.expo, TimeoutError, max_tries=10)
+def get_one_thread(service: str, user_id: str, thread_id: str) -> pd.DataFrame:
+    imported_timestamp = get_current_timestamp()
+    thread_results = service.users().threads().get(userId=user_id, id=thread_id).execute()
+
+    df_thread = pd.DataFrame()
+    messages = thread_results['messages']
+
+    df_thread['threadId'] = [thread_results['id']]
+    df_thread['historyId'] = [thread_results['historyId']]
+    df_thread['total_email_count'] = [len(messages)]
+
+    # to find count of emails from a specific address in a thread (sent_email_count)
+    sent_email_count = 0
+    email_sent_pattern = 'elifesciences.org'
+    for message in messages:
+        header_list = message['payload']['headers']
+        for header in header_list:
+            if header['name'] == 'From':
+                if email_sent_pattern in header['value']:
+                    sent_email_count += 1
+
+    df_thread['sent_email_count'] = sent_email_count
+
+    # to get first responder
+    # pylint: disable=consider-using-enumerate
+    first_response_index = 0
+    for i in range(len(messages)):
+        header_list = messages[i]['payload']['headers']
+        for header in header_list:
+            if header['name'] == 'From':
+                if email_sent_pattern in header['value']:
+                    df_thread['first_responder'] = [header['value']]
+                    first_response_index = i
+                    break
+
+    # to get first response date
+    header_list = messages[first_response_index]['payload']['headers']
+    for header in header_list:
+        if header['name'] == 'Date':
+            df_thread['first_reponse_date'] = [header['value']]
+            break
+
+    df_thread['user_id'] = user_id
+    df_thread['imported_timestamp'] = imported_timestamp
+
+    return df_thread
+
+
 def write_dataframe_to_jsonl_file(
         df_data_to_write: pd.DataFrame,
         target_file_path: str):
 
     with open(target_file_path, 'w', encoding='utf-8') as file:
         # newline-delimited JSON
         df_data_to_write.to_json(file, orient='records', lines=True, force_ascii=False)
+
+
+@backoff.on_exception(backoff.expo, TimeoutError, max_tries=10)
+def get_distinct_values_from_bq(
+            project_name: str,
+            dataset: str,
+            column_name: str,
+            table_name: str
+        ) -> pd.DataFrame:
+
+    sql = (
+        """
+        SELECT DISTINCT {}
+        FROM  `{}.{}.{}`
+        """.format(column_name, project_name, dataset, table_name)
+    )
+
+    client = bigquery.Client(project_name)
+    query_job = client.query(sql)  # API request
+    results = query_job.result()  # Waits for query to finish
+
+    df_value = pd.DataFrame()
```

Comment:
I am not 100% sure what this is doing. Is it generating a DataFrame with a single column (`column_name`) where the value comes from `row.threadId`?

If so, this would perhaps be shorter?

```python
df_value = pd.DataFrame({
    {content_name: row.threadId}
} for row in results)
```

But you could also consder using pandas' [read_gbq](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_gbq.html). (I am using it in [data-science-dags](https://github.com/elifesciences/data-science-dags/blob/46559fa41d8641137981d1490466e4e7363d0a3a/data_science_pipeline/utils/jupyter.py#L18), but that probably doesn't help much because it's called by notebooks).

---

As a side note, when I suggested [pd.concat](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.concat.html), I think it will only really make things faster if you call it once (not per row). e.g.:

```python
pd.concat([
    some_method_returning_a_dataframe(item)
    for item in items
])
```

e.g. if `items` was a list of thread ids and `some_method_returning_a_dataframe` was a function that returned a DataFrame containing messages for that thread id.

Reply:
Thank you for mentioning about this usage of `concat`!